java/ql/lib/semmle/code/java/regex/regex.qll

```diff
@@ -0,0 +1,933 @@
+      this.quoteDelimiter(inner_end) = false and
+      not exists(int mid |
+        this.quoteDelimiter(mid) = false and mid in [inner_start .. inner_end - 1]
+      )
+    )
+  }
+
+  /**
+   * A control sequence, `\cx`
+   * `x` may be any ascii character including special characters.
+   */
+  predicate controlEscape(int start, int end) {
+    this.escapingChar(start) and
+    this.getChar(start + 1) = "c" and
+    end = start + 3
+  }
+
+  /** Gets the text of this regex */
+  string getText() { result = this.(StringLiteral).getValue() }
+
+  string getChar(int i) { result = this.getText().charAt(i) }
+
+  string nonEscapedCharAt(int i) {
+    result = this.getText().charAt(i) and
+    not exists(int x, int y | this.escapedCharacter(x, y) and i in [x .. y - 1]) and
+    not exists(int x, int y | this.quote(x, y) and i in [x .. y - 1]) and
+    not exists(int x, int y | this.controlEscape(x, y) and i in [x .. y - 1])
+  }
+
+  private predicate isOptionDivider(int i) { this.nonEscapedCharAt(i) = "|" }
+
+  private predicate isGroupEnd(int i) { this.nonEscapedCharAt(i) = ")" and not this.inCharSet(i) }
+
+  private predicate isGroupStart(int i) { this.nonEscapedCharAt(i) = "(" and not this.inCharSet(i) }
+
+  predicate failedToParse(int i) {
+    exists(this.getChar(i)) and
+    not exists(int start, int end |
+      this.topLevel(start, end) and
+      start <= i and
+      end > i
+    )
+  }
+
+  /** An escape sequence that includes braces, such as named characters (\N{degree sign}), named classes (\p{Lower}), or hex values (\x{h..h}) */
+  private predicate escapedBraces(int start, int end) {
+    this.escapingChar(start) and
+    this.getChar(start + 1) = ["N", "p", "P", "x"] and
+    this.getChar(start + 2) = "{" and
+    this.getChar(end - 1) = "}" and
+    end > start and
+    not exists(int i | start + 2 < i and i < end - 1 | this.getChar(i) = "}")
+  }
+
+  /**
+   * Holds if an escaped character is found between `start` and `end`.
+   * Escaped characters include hex values, octal values and named escapes,
+   * but excludes backreferences.
+   */
+  predicate escapedCharacter(int start, int end) {
+    this.escapingChar(start) and
+    not this.backreference(start, _) and
+    (
+      // hex value \xhh
+      this.getChar(start + 1) = "x" and
+      this.getChar(start + 2) != "{" and
+      end = start + 4
+      or
+      // octal value \0o, \0oo, or \0ooo. Max of 0377.
+      this.getChar(start + 1) = "0" and
+      end in [start + 3 .. start + 5] and
+      forall(int i | i in [start + 1 .. end - 1] | this.isOctal(i)) and
+      (end = start + 5 implies this.getChar(start + 2) <= "3") and
+      not (
+        end < start + 5 and
+        this.isOctal(end) and
+        (end = start + 4 implies this.getChar(start + 2) <= "3")
+      )
+      or
+      // 16-bit hex value \uhhhh
+      this.getChar(start + 1) = "u" and end = start + 6
+      or
+      this.escapedBraces(start, end)
+      or
+      // Boundry matchers \b, \b{g}
```

Comment:
I notice you're not handling the other forms of boundary matchers I see in [the docs](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html), such as `\A`, `\Z`, and `\G`.

We missed those originally in Ruby as well, and they caused weird messages from the ReDoS query alerts, where it thought for example that `\G` was just an escape of the character `G`, so it suggested the regex `\G(a*)*b` would backtrack exponentially for strings of the form `Gaaaaaa...`. I fixed that in https://github.com/github/codeql/commit/f63c768d9f48db8d57604f290dc442876b6812ff.